### PR TITLE
[rllib] Fix missing policy learner_stats in multi-gpu learners

### LINKED ISSUE #18116: [rllib] Policy `learner_stats` get dropped when multi_gpu_learner_thread.py is used (in GPU and multi-GPU use cases).
### What is the problem?

When multiple gpus are used, learner stats are gathered in the `learn_on_loaded_batch` method with a "tower_X" key before the stats:
https://github.com/ray-project/ray/blob/master/rllib/policy/torch_policy.py#L645

for example:
```
{'tower_0': {'learner_stats': {'cur_lr': 0.000495184, 'policy_loss': -40.517921447753906, 'entropy': 1.8380180597305298, 'entropy_coeff': 0.0005, 'var_gnorm': 17.741676330566406, 'vf_loss': 0.7066917419433594, 'vf_explained_var': array([0.5089742], dtype=float32), 'mean_rhos': 1.0025060176849365, 'std_rhos': 0.39850014448165894}}}
```

This 'tower_0' is not taken into account when `get_learner_stats()` is used:
https://github.com/ray-project/ray/blob/089dd9b94924cccb85ba1affccc4b0c8907f192c/rllib/execution/multi_gpu_learner_thread.py#L102

This causes the policy learner_stats to get dropped when GPUs are used.

## CPU does not drop these stats
This is different from when a single gpu/cpu is used, the `learn_on_loaded_batch` function will return:
```
{'learner_stats': {'cur_lr': 0.000495184, 'policy_loss': -40.517921447753906, 'entropy': 1.8380180597305298, 'entropy_coeff': 0.0005, 'var_gnorm': 17.741676330566406, 'vf_loss': 0.7066917419433594, 'vf_explained_var': array([0.5089742], dtype=float32), 'mean_rhos': 1.0025060176849365, 'std_rhos': 0.39850014448165894}
```

(note the lack of `tower_X` key)

Similar code can then extract the policy metrics which works!
https://github.com/ray-project/ray/blob/089dd9b94924cccb85ba1affccc4b0c8907f192c/rllib/execution/learner_thread.py#L80

*Ray version and other system information (Python version, TensorFlow version, OS):*

version: latest dev 2.0.0
python: 3.8
macosx + linux
torch + tensorflow

### Reproduction (REQUIRED)

Run anything with GPU learners (specifically in my case I'm using IMPALA)

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

@sven1977

### PR DESCRIPTION
**Pull request by Bam4d**
@sven1977 

## Why are these changes needed?

Details of bug are outlined in #18116 and the fix suggested by @mvindiola1 is implemented.

## Related issue number

Closes #18116

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/torch_policy.py [hunk 1]
```diff
@@ -641,12 +641,11 @@ def learn_on_loaded_batch(self, offset: int = 0, buffer_index: int = 0):
...
         batch_fetches = {}
+        batch_fetches[LEARNER_STATS_KEY] = {}
         for i, batch in enumerate(device_batches):
-            batch_fetches[f"tower_{i}"] = {
-                LEARNER_STATS_KEY: self.extra_grad_info(batch)
-            }
+            batch_fetches[LEARNER_STATS_KEY][f"tower_{i}"] = self.extra_grad_info(batch)
 
-        batch_fetches.update(self.extra_compute_grad_fetches())
+        batch_fetches[LEARNER_STATS_KEY].update(self.extra_compute_grad_fetches())
```
Comment (mvindiola1) [inline rllib/policy/torch_policy.py:648]: @Bam4d,
Nice. What do you think of using merge_dicts which will merge them recursively.

https://github.com/ray-project/ray/blob/d9552e67950c0211d727a11873a5c95dd510108f/python/ray/util/ml_utils/dict.py#L9-L20

```python
from ray.rllib.utils import merge_dicts

batch_fetches = merge_dicts(batch_fetches, self.extra_compute_grad_fetches(), True)
```

Comment (Bam4d) [reply rllib/policy/torch_policy.py:648]: I think this is probably a better way of doing it thanks!

**Comment by Bam4d** [comment]
So this fix breaks a few other algorithms such as PPO where PPO is specifically looking for the losses in the `LEARNER_STATS_KEY`

For example in UpdateKL class:
```
class UpdateKL:
    """Callback to update the KL based on optimization info.

    This is used inside the execution_plan function. The Policy must define
    a `update_kl` method for this to work. This is achieved for PPO via a
    Policy mixin class (which adds the `update_kl` method),
    defined in ppo_[tf|torch]_policy.py.
    """

    def __init__(self, workers):
        self.workers = workers

    def __call__(self, fetches):
        def update(pi, pi_id):
            assert LEARNER_STATS_KEY not in fetches, \
                ("{} should be nested under policy id key".format(
                    LEARNER_STATS_KEY), fetches)
            if pi_id in fetches:
                kl = fetches[pi_id][LEARNER_STATS_KEY].get("kl")
                assert kl is not None, (fetches, pi_id)
                # Make the actual `Policy.update_kl()` call.
                pi.update_kl(kl)
            else:
                logger.warning("No data for {}, not updating kl".format(pi_id))

        # Update KL on all trainable policies within the local (trainer)
        # Worker.
        self.workers.local_worker().foreach_trainable_policy(update)
```

Not really sure how to fix this especially if PPO is split across multiple towers (is this possible)? because the KL will have to be calculated per-power and applied per-tower right?

**Comment by mvindiola1** [comment]
@Bam4d

Question for you. Are the stats values actually different between the towers? 

Looking at the code here I would think they would all be the same since they are obtained after the update. https://github.com/ray-project/ray/blob/97f71e15d48d13b12a894f2721a2eb40adcb0a6e/rllib/policy/torch_policy.py#L646

If that is true there is a race condition here because each thread calculating the loss is storing the stats in the same policy.statskey member variables. I guess the policy will, in the best case, have the stats of the last thread to finish.

Even if that is sorted out there is still the issue of combining them you brought up.

**Comment by Bam4d** [comment]
The problem I can see with averaging is that those stats are getting used in part of the calculation in PPO so only using them from a single tower or averaging them might break the algorithm completely.

I'm pretty sure there are many places that would break differently depending on what the fix is....

Not sure how to proceed really...

**Comment by Bam4d** [comment]
For example in Vtrace only tower 0 actually records stats. 
https://github.com/ray-project/ray/blob/master/rllib/agents/impala/vtrace_torch_policy.py#L186

In this specific case these values are used for stats, but in other algorithms (like the PPO example) if the training is on multiple towers, then different stats may need to be summed or averaged depending on what the stats are. These collated stats from multiple towers would then be used in the KL calculation here for example.

**Comment by sven1977** [comment]
Hey guys, thanks for looking into this! True, we have to be careful with these stats as some of them may be essential for learning (td-errors for prio replay or kl in case of PPO). Also true that there is a race condition here. The cleanest way would probably be (for torch algos) to store these stats per model (==tower) and then average them (except for td-error, which needs to stay per-batch-item). I remember we are already doing something similar for the td-errors in SAC/DQN/etc..

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.
